Feature/fixed width parsing

blrnw3 opened the pull request:
Needs some more tests but otherwise working well. Currently using a JAR for the modified databricks-spark-csv code but hoping my changes will get merged (https://github.com/databricks/spark-csv/pull/259), else I'll change it a git dependency on our fork of databricks (https://github.com/quartethealth/spark-csv).

Commit:
Fixed-width parsing spark integration.
Based on and heavily reliant on databricks-spark-csv (a slightly modified version for now)
Initial commit with basic tests. TODO: more tests

Commit:
Added docs. Some renaming

blrnw3 commented:
Again, @pheuter @thyming, not sure if you want to review this or not but please feel free to criticize as much as you like ;D

thyming commented:
I will review upon my return from Puerto Rico, thanks!
On Thu, Feb 11, 2016 at 6:49 PM Ben LR <notifications@github.com> wrote:

> Again, @pheuter <https://github.com/pheuter> @thyming
> <https://github.com/thyming>, not sure if you want to review this or not
> but please feel free to criticize as much as you like ;D
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/quartethealth/spark-fixedwidth/pull/1#issuecomment-183113930>
> .
>

Commit:
More tests and readme
Also made the whitespace options actually useable (databricks spark-csv exposes but ignores these).

Commit:
Readme and code style fixes

Commit:
Added dependency on our github-hosted fork of spark-csv
This means we don't need to use a JAR for the dep.

thyming commented on src/test/scala/com/quartethealth/spark/fixedwidth/FixedwidthSuite.scala:
```diff
@@ -0,0 +1,117 @@
+    StructField("cost", DoubleType)
+  ))
+
+  protected var sqlContext: SQLContext = _
```
you should be able to make this an immutable `lazy val` with the expression that you execute below, and it'll evaluate on first usage then not be modified otherwise.

blrnw3 replied:
Come to think of it, I'm not convinced we need a new `sqlContext` every time, so I've turned it into a straight `val`

thyming commented on src/test/scala/com/quartethealth/spark/fixedwidth/FixedwidthSuite.scala:
```diff
@@ -0,0 +1,117 @@
+    val result = sqlContext.fixedFile(fruit_resource("malformed"), fruitWidths,
+      useHeader = false, mode = "DROPMALFORMED")
+    result.show()
+    assert(result.collect().length < fruitSize)
```
can this be more specific than `<`?

thyming commented on src/test/scala/com/quartethealth/spark/fixedwidth/FixedwidthSuite.scala:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  test("Parse malformed, with schema, FAILFAST") {
+    intercept[Exception](
```
Catch a more specific exception? I see the readme says runtimeexception, is spark-csv using somethign more fine-grained than that? even so, to me that's better than general `Exception`

blrnw3 replied:
Yeah, I too thought it should throw a `RuntimeException`. In fact it does, but it then throws a `SparkException` and the test fails to capture the inner `RunTimeException`. I guess because Spark acts like a container for the test.  But you are right, anything is better than a generic `Exception`

thyming commented:
Very nice overall!
One thing I think we should just agree on as a team is which test framework to use - @pheuter and I had started writing tests with https://etorreborre.github.io/specs2/ , my bad for not communicating that to you. If you have a strong preference for scalatest then make the argument for it - otherwise I think we should choose a single one and stick to it. I'm happy to help migrate these tests.

thyming commented on src/test/scala/com/quartethealth/spark/fixedwidth/FixedwidthSuite.scala:
```diff
@@ -0,0 +1,117 @@
+        fruitSchema, useHeader = false, mode = "FAILFAST").collect()
+    )
+  }
+
```
Is there a test that covers a well-formed file but the incorrect schema? My sense is that is likely, but just want to make sure.

blrnw3 replied:
Good spot. I'll add that. Again, I'll only be able to catch it throwing a `SparkException`, but this is better than no test at all.

thyming commented on src/main/scala/com/quartethealth/spark/fixedwidth/FixedwidthRelation.scala:
```diff
@@ -0,0 +1,56 @@
+    inferSchema: Boolean,
+    codec: String = null,
+    nullValue: String = "")(@transient override val sqlContext: SQLContext)
+  extends CsvRelation(
```
How does it work that this is extending the csv parser? I see code like this https://github.com/quartethealth/spark-csv/blob/master/src/main/scala/com/databricks/spark/csv/CsvRelation.scala#L80-L86 and wonder if extension is the right choice here. At the very least it seems sort of unintuitive to me that a fixed-width parser is a case of a csv parser.

blrnw3 replied:
You're right, it doesn't make a lot of sense from an ideal design perspective. However I wanted to make as few changes as possible to the original source, so this was a compromise. I'd be willing to be persuaded otherwise though...

blrnw3 commented:
Thanks again for the comments Luke. All seem very sensible.
As for the test framework, I was just using what spark-csv used. I have no strong preference and I see no reason why we shouldn't use a different one from what they use. So I'll happily migrate these to spec format.

pheuter commented:
seems to me like a lot of the feedback in this PR is around tooling and structures that were already in place from the original spark-csv repo this was forked from. a question i have is where do we stand as a team on spending work refactoring code we didn't create to conform to our conventions? since we technically own this code, it makes sense to conform to our standards, but at the same time, keeping the fork as close as possible to the base so it can pull upstream changes has its benefits as well.

blrnw3 commented:
My main concern is indeed having to deal with merge conflicts when we want to take advantage of any updates to the upstream code. But for anything which wouldn't be affected, like tests, conforming to our standards seems the best policy.

thyming commented:
Well this isn't a fork, it's an extension, so I'm not sure what you're
saying applies in this case, Mark

On Wed, Feb 17, 2016 at 11:16 AM Mark Fayngersh <notifications@github.com>
wrote:

> seems to me like a lot of the feedback in this PR is around tooling and
> structures that were already in place from the original spark-csv repo this
> was forked from. a question i have is where do we stand as a team on
> spending work refactoring code we didn't create to conform to our
> conventions? since we technically own this code, it makes sense to conform
> to our standards, but at the same time, keeping the fork as close as
> possible to the base so it can pull upstream changes has its benefits as
> well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/quartethealth/spark-fixedwidth/pull/1#issuecomment-185278935>
> .
>

Commit:
Convert tests to specs2 format
Tidied-up a few things as per the PR comments

blrnw3 commented:
@pheuter  @thyming  Let me know if there are any outstanding issues with this otherwise I'll merge it tomorrow. Thanks

thyming commented:
Cool, looks good
